### PR TITLE
Fix remaining feature gaps in DfaMatcher

### PR DESCRIPTION
**Pull request by rynowak**
/cc @kichalla @JamesNK 

Here's a DFA matcher that works with MatchProcessor and EndpointConstraint

### CODE REVIEW

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherAzureBenchmark.cs [hunk 1]
```diff
@@ -30,7 +30,7 @@ public void Setup()
             _samples = SampleRequests(EndpointCount, SampleCount);
 
             _baseline = (BarebonesMatcher)SetupMatcher(new BarebonesMatcherBuilder());
-            _dfa = SetupMatcher(new DfaMatcherBuilder());
+            _dfa = SetupMatcher(CreateDfaMatcherBuilder());
```
Comment (rynowak) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/Matchers/MatcherAzureBenchmark.cs]: A lot of the churn here is due to adding some services to the DFA matcher's constructor.

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/Microsoft.AspNetCore.Routing.Performance.csproj [hunk 2]
```diff
@@ -18,23 +18,12 @@
...
     <Compile Include="..\..\test\Microsoft.AspNetCore.Routing.Tests\Matchers\BarebonesMatcherBuilder.cs">
       <Link>Matchers\BarebonesMatcherBuilder.cs</Link>
     </Compile>
-    <Compile Include="..\..\test\Microsoft.AspNetCore.Routing.Tests\Matchers\DfaMatcher.cs">
-      <Link>Matchers\DfaMatcher.cs</Link>
-    </Compile>
-    <Compile Include="..\..\test\Microsoft.AspNetCore.Routing.Tests\Matchers\DfaMatcherBuilder.cs">
-      <Link>Matchers\DfaMatcherBuilder.cs</Link>
-    </Compile>
-    <Compile Include="..\..\test\Microsoft.AspNetCore.Routing.Tests\Matchers\MatcherBuilderEntry.cs" Link="Matchers\MatcherBuilderEntry.cs" />
```
Comment (rynowak) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/Microsoft.AspNetCore.Routing.Performance.csproj]: These have all been moved into product code

File: src/Microsoft.AspNetCore.Routing.Abstractions/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -3,4 +3,5 @@
...
\ No newline at end of file
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.Mvc.Routing.Abstractions.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing.Abstractions/Properties/AssemblyInfo.cs]: I noticed we were missing these, and I'm using Moq in one place so I went ahead and added them

File: src/Microsoft.AspNetCore.Routing/Matchers/CandidateSet.cs [hunk 4]
```diff
@@ -39,6 +40,20 @@ public CandidateSet(Candidate[] candidates, int[] groups)
...
+
+        public IEnumerator<MatcherEndpoint> GetEnumerator()
+        {
+            return Candidates.Select(c => c.Endpoint).GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Matchers/CandidateSet.cs]: Will remove the `IReadOnlyList<>` part of this. It's unrelated to this PR.

File: src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs [hunk 5]
```diff
@@ -161,7 +161,7 @@ public bool TryMatch(PathString path, RouteValueDictionary values)
                 }
                 if (!pathSegment.IsSimple)
                 {
-                    if (!MatchComplexSegment(pathSegment, requestSegment.ToString(), Defaults, values))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs]: `Defaults` on this code path was totally unused.

File: src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs [hunk 6]
```diff
@@ -287,10 +287,9 @@ private bool SavePathSegmentsAsValues(int index, RouteValueDictionary values, St
             return false;
         }
 
-        private bool MatchComplexSegment(
+        internal static bool MatchComplexSegment(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs]: This is the existing code for processing complex segments. If we need to at some point we can work on the perf of it (I'm sure it's bad). As it is though it just works with the DFA matcher

File: src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs [hunk 7]
```diff
@@ -287,10 +287,9 @@ private bool SavePathSegmentsAsValues(int index, RouteValueDictionary values, St
...
             RoutePatternPathSegment routeSegment,
             string requestSegment,
-            IReadOnlyDictionary<string, object> defaults,
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs]: Totally unused

File: src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs [hunk 8]
```diff
@@ -499,7 +496,7 @@ private bool SavePathSegmentsAsValues(int index, RouteValueDictionary values, St
             {
                 foreach (var item in outValues)
                 {
-                    values.Add(item.Key, item.Value);
+                    values[item.Key] = item.Value;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs]: This is a quirk of how 'slots' for route values work. You sometimes need to overwrite a value. This won't have any functional impact on the old code path.

**Comment by rynowak** [comment]
@kichalla @JamesNK this is on the shortest path to making the DfaMatcher the default 😆

File: src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+            for (var i = depth; i < entry.Pattern.Segments.Count; i++)
+            {
+                var segment = entry.Pattern.Segments[i];
+                if (!segment.IsSimple)
+                {
+                    // Complex segments always require more processing
+                    return true;
+                }
+
+                if (segment.Parts[0].IsLiteral)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs]: Assign Parts[0] to a variable

File: test/Microsoft.AspNetCore.Routing.Tests/Matchers/FullFeaturedMatcherConformanceTest.cs [hunk 10]
```diff
@@ -185,6 +189,9 @@ public abstract class FullFeaturedMatcherConformanceTest : MatcherConformanceTes
 
             // Assert
             DispatcherAssert.AssertNotMatch(feature);
+
+            GC.KeepAlive(keys);
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Routing.Tests/Matchers/FullFeaturedMatcherConformanceTest.cs]: Hmmm? What are these for? Add a comment

File: src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcher.cs [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+                var start = groups[i];
+                var length = groups[i + 1] - groups[i];
+                var group = candidatesArray.AsSpan(start, length);
+
+                // Yes, these allocate. We should revise how this interaction works exactly
+                // once the extensibility is more locked down.
+                //
+                // Would could produce a fast path for a small number of members in
+                // a group.
+                var members = new BitArray(group.Length);
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcher.cs]: I don't want to go down that rabbit hole in this PR. This is going to keep changing based on the details we discussed yesterday.

File: src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+            }
+        }
+
+        // internal for tests
+        internal Candidate CreateCandidate(MatcherBuilderEntry entry)
+        {
+            var assignments = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+            var slots = new List<KeyValuePair<string, object>>();
+            var captures = new List<(string parameterName, int segment, int index)>();
+            (string parameterName, int segment, int index) catchAll = default;
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs]: I did a pass to change a bunch of these to `segmentIndex`/`slotIndex`. If you find more after I'm done here then point them out, but hopefully I found them all.

**Pull request merged**